### PR TITLE
Index format tests

### PR DESCRIPTION
**Pull request by haoxli**
Add test cases for testing drawIndexed, IndexFormat, Primitive restart

### CODE REVIEW

**Comment by kainino0x** [comment]
Since this uses GLSL, it should be a pull request against the `glsl-dependent` branch. Could you rebase it and open a new PR?

**Comment by haoxli** [comment]
OK, close this one.

File: src/webgpu/api/operation/command_buffer/render/index_format.spec.ts [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+    });
+  }
+
+  MakeBufferMapped(arrayBuffer: ArrayBuffer, usage: number): GPUBuffer {
+    const [buffer, bufferMapping] = this.device.createBufferMapped({
+      size: arrayBuffer.byteLength,
+      usage,
+    });
+
+    if (arrayBuffer instanceof Float32Array) {
```
Comment (kainino0x) [inline src/webgpu/api/operation/command_buffer/render/index_format.spec.ts:115]: Won't these always be false?

File: src/webgpu/api/operation/command_buffer/render/index_format.spec.ts [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+    this.expectContents(this.result, expected);
+  }
+}
+
+export const g = new TestGroup(IndexFormatTest);
+
+g.test('uint32').fn(async t => {
+  await t.initResources('uint32');
+
+  const vertexArray = new Float32Array([
```
Comment (kainino0x) [inline src/webgpu/api/operation/command_buffer/render/index_format.spec.ts:164]: Consider putting /* prettier-ignore */  on this so it doesn't autoformat so poorly

File: src/webgpu/api/operation/command_buffer/render/index_format.spec.ts [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+  private swapChainFormat: GPUTextureFormat = 'bgra8unorm';
+
+  async initResources(format: GPUIndexFormat): Promise<void> {
+    // glslang module
+    this.glslang = await glslangModule();
+
+    // render pipeline
+    this.pipeline = this.MakeRenderPipeline(format);
+
+    const context = this.CreateRenderContext(true);
```
Comment (kainino0x) [inline src/webgpu/api/operation/command_buffer/render/index_format.spec.ts:21]: Don't use a swap chain to get a render target; just use a regular createTexture. Take a look at some other tests on the glsl-dependent branch which test rendering results.

File: src/webgpu/api/operation/command_buffer/render/index_format.spec.ts (hunk 1 above)
Comment (haoxli) [reply src/webgpu/api/operation/command_buffer/render/index_format.spec.ts:115]: I tested it's true if arrayBuffer = new Float32Array()

Comment (kainino0x) [reply src/webgpu/api/operation/command_buffer/render/index_format.spec.ts:115]: Actually, it's just the type annotation of `arrayBuffer` looks wrong.

I think you should remove this helper and inline the code where you use it now; the inline sites should be quite small.

File: src/webgpu/api/operation/command_buffer/render/index_format.spec.ts (hunk 3 above)
Comment (haoxli) [reply src/webgpu/api/operation/command_buffer/render/index_format.spec.ts:21]: Thanks, I will use the way in glsl-dependent branch.

**Pull request closed without merging**